Thorlabs power meter support based on TLPM driver (closes #39)

qku opened the pull request:
<!--- Provide a general short and descriptive title above -->

## Description
<!--- Describe your changes in detail -->
Add a new hardware file for the Thorlabs power meter based on `process_control_interface`. Requires the Optical Power Monitor software to be installed on the computer. The currently measured power can be read out as a process value. The measurement wavelength can be as a config option.

Error messages with useful hints are shown if the DLL is not installed or the power meter is likely set to work with the legacy PM100D driver (the latter can be changed with the Power Meter Driver Switcher application installed along the Optical Power Monitor).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
As mentioned in issue #39, the old implementation based on the PM100D driver has not yet been made compatible with the new core. Since the PM100D driver is legacy and all old power meters should also be supported by the TLPM driver, a new implementation based on TLPM is sensible.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Using the example config (with the correct address) and a PM100A power meter. Missing DLL and incorrect driver setting (to PM100D driver) was also simulated to test the error messages.

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Put an 'x' in all the boxes that apply, e.g. "[x]". -->
<!--- You can also create this PR and afterwards check the boxes by clicking on them. -->
- [X] Bug fix
- [X] New feature
- [ ] Breaking change (Causes existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an 'x' in all the boxes that apply, e.g. "[x]". -->
<!--- If you're unsure about any of these, ask. -->
<!--- You can also create this PR and afterwards check the boxes by clicking on them. -->
- [X] My code follows the code style of this project.
- [X] I have documented my changes in `/docs/changelog.md`.
- [ ] My change requires additional/updated documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added/updated the config example for any module docstrings as necessary.
- [X] I have checked that the change does not contain obvious errors
(syntax, indentation, mutable default values, etc.).
- [ ] I have tested my changes using 'Load all modules' on the default dummy configuration.
- [ ] All changed Jupyter notebooks have been stripped of their output cells.

GRexBayer commented:
Hi, 
I tested the hardware file and it seems to work fine. However I have two points:
1. If I am not mistaken, the default config should use the path "module.Class: 'powermeter.thorlabs_powermeter.ThorlabsPowermeter'" instead.
2.  I wonder if there is a simple way to find out the address of the powermeter and if we could somehow include some documentation about this in the file such that people find it out more easily. I believe remember the old software for the powermeter showed it upon loading the PM. Now, I could not find this part in the new Thorlabs software. I guessed the new address with help of the device instance path in the device manager. I am quite sure there is a better way

GRexBayer commented:
> Hi, I tested the hardware file and it seems to work fine. However I have two points:
> 
> 1. If I am not mistaken, the default config should use the path "module.Class: 'powermeter.thorlabs_powermeter.ThorlabsPowermeter'" instead.
> 2. I wonder if there is a simple way to find out the address of the powermeter and if we could somehow include some documentation about this in the file such that people find it out more easily. I believe remember the old software for the powermeter showed it upon loading the PM. Now, I could not find this part in the new Thorlabs software. I guessed the new address with help of the device instance path in the device manager. I am quite sure there is a better way

Ah, also: 
It was tested on win 10 with a PM100D with Firmware 2.4.0
Application 4.1.4204.742
Driver TLPMX__32 5.3.4204.576

LukePiewalker reviewed (approved):
Looks good to me. But I do have some suggestions regarding the activity state of the hardware. These are not really crucial, so therefore I would approve as is in case these suggestions are not to be considered.

I tested with a PM100D and a PM16-120 on an office PC with the "default.cfg" extended with the config from the docstring (with the correct USB adress of course)

qku commented:
Hi @GRexBayer,
thank you for testing the module.

> 1. If I am not mistaken, the default config should use the path "module.Class: 'powermeter.thorlabs_powermeter.ThorlabsPowermeter'" instead.

That is true, I fixed this with 9a21d53.

> 2. I wonder if there is a simple way to find out the address of the powermeter and if we could somehow include some documentation about this in the file such that people find it out more easily. I believe remember the old software for the powermeter showed it upon loading the PM. Now, I could not find this part in the new Thorlabs software. I guessed the new address with help of the device instance path in the device manager. I am quite sure there is a better way

The module gets a list of all available powermeter (addresses) from the DLL during module activation. There was already an info-level log message with those, but I guess it is easily overlooked. I added a comment to the example config.

I also made the address config option optional. The default behaviour in this case is to use the first powermeter in the list. My thought was that this could help people to get started quickly. It might also lead to unexpected results though, if people just forget to provide the address. Do you think this behaviour is good?

qku commented on src/qudi/hardware/powermeter/thorlabs_powermeter.py:
```diff
@@ -0,0 +1,228 @@
+        """
+        return self._constraints
+
+    def set_activity_state(self, channel, active):
```
That's a great idea. I created separate (hidden) methods to initialize and close the powermeter. These are now called by `set_activity_state()` and the activity state is checked before process values are given out.

qku commented:
@GRexBayer after some thinking, I believe that going with the first powermeter in the list is not such a good idea. The risk of reading values from the device might just be too high to justify the simplified startup. Let me know what you think!

GRexBayer commented:
> @GRexBayer after some thinking, I believe that going with the first powermeter in the list is not such a good idea. The risk of reading values from the device might just be too high to justify the simplified startup. Let me know what you think!

you mean that it might read values that actually should not be read from the wrong device which might lead to complications with the logic?
My thought until was that it is more convenient that it picks the PM automatically. And since the default config also contains a random address of the PM, it would be a conscious decision to comment it out and let QuDi handle the recognition. Also, if you use two PMs, you should understand that QuDi needs to know which one of the two. In the software meeting, there seemed to be no veto against it.
Maybe if you know a scenario where it might cause harm, we can omit this

Neverhorst commented:
It is quite common practice to automatically connect to the first found device if no specific device is given. As long as you can ensure the device is actually of the type you want to connect to (right model/make), there are no weird things happening. And since the Thorlabs driver only lists compatible powermeters you should be safe.
My recommendation is to take the first device by default if no respective `ConfigOption` is given, but set `missing='warn'` in order to warn the user about the missing configuration and what is happening.

qku commented:
If this is common practice to connect to the first found device, I would also do that here. I was mostly thinking about accidentally reading values from the wrong powermeter in the lab (say the one connected to laser 2 while you want to know the power behind laser 1). But I guess this would usually be easily detected by the user, who has at least an expectation of the order of magnitude.

The current state implements @Neverhorst suggestion with `missing='warn'`. This makes the PR ready to merge from my side.

GRexBayer reviewed (approved):
The file was tested on Windows 10 with the new core in combination with the PID interface. The requested changes have been implemented and the module runs as desired.
